api/src/main/java/org/openmrs/layout/name/NameSupport.java

```diff
@@ -27,16 +38,100 @@ public NameSupport() {
 	}
 	
 	public static NameSupport getInstance() {
-		if (singleton == null) {
-			throw new APIException("Not Yet Instantiated");
-		} else {
-			return singleton;
+		synchronized (NameSupport.class) {
+			singleton.init();
+			if (singleton == null) {
+				throw new APIException("Not Yet Instantiated");
+			} else {
+				return singleton;
+			}
 		}
 	}
+
+	/**
+	 * Initializes layout templates with a custom template configured
+	 * via the "layout.name.template" GP.
+	 */
+	private void init() {
+		if (initialized) {
+			return;
+		}
+		Context.getAdministrationService().addGlobalPropertyListener(singleton);
+		// Get configured name template to override the existing one if any
+		String layoutTemplateXml = Context.getAdministrationService().getGlobalProperty(
+			OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_TEMPLATE);
+		NameTemplate nameTemplate = deserializeXmlTemplate(layoutTemplateXml);
+		
+		if (nameTemplate != null) {
+			updateLayoutTemplates(nameTemplate);
+			initialized = true;
+		}
+	}
+	
+	/**
+	 * Update existing layout templates if present with the provided template
+	 */
+	private void updateLayoutTemplates(NameTemplate nameTemplate) {
+		if (getLayoutTemplates() == null) {
+			setLayoutTemplates(new ArrayList<>());
+		}
+		List<NameTemplate> list = new ArrayList<>();
+		// filter out unaffected templates to keep
+		list.addAll(getLayoutTemplates().stream().filter(existingTemplate -> existingTemplate.getCodeName() != nameTemplate.getCodeName()).collect(Collectors.toList()));
+		list.add(nameTemplate);
+		setLayoutTemplates(list);
+	}
 	
+	/**
+	 * @return Returns the defaultLayoutFormat
+	 */
+	private NameTemplate deserializeXmlTemplate(String xml) {
+		NameTemplate nameTemplate = null;
+		try {
+			nameTemplate = Context.getSerializationService().getDefaultSerializer().deserialize(xml,
+				NameTemplate.class);
+		} catch (Exception e) {
+			log.error("Error in deserializing provided name template", e);
```

Comment:
Also here, a null or empty GP probably doesn't need to be treated as an error.

Reply:
Addressed. Method now returns early for all null and empty strings provided.